File: analysis/calc.pl
use DBI;
use strict;
use warnings;

STDOUT->autoflush(1);

my $dbfile = "/home/elimtob/Workspace/mymemtrace/traces/imagick_r-test.db";

my $dbh = DBI->connect("dbi:SQLite:dbname=$dbfile","","") || die "Noope";

my $stmt = $dbh->prepare("SELECT SUM(REFS) from ROW_COUNT");
$stmt->execute();
my $ans = $stmt->fetch;
my $rows = $$ans[0];
print("Total number of rows: $rows\n");

$stmt = $dbh->prepare("select ADDR >> 6 from MEMREFS;");
$stmt->execute();

my %hist   = ();
my $kmin   = 0;
my $max    = 0;
my $count  = 0;
my $gcount = 0;
$hist{$kmin} = $rows;

while (my $ans = $stmt->fetch) {
    my $cl = $$ans[0];
    if (exists($hist{$cl})) {
        $hist{$cl}++;
    } else {
        $hist{$cl} = 1;
        $count++;
    }
    $max = $hist{$cl} unless $max >= $hist{$cl};
    $kmin = $cl unless $hist{$kmin} <= $hist{$cl};
    $gcount++;
    printf(" %6.2f%% done\r", $gcount / $rows * 100) if $gcount % 10000 == 0;
}
print("\n");
print("Max = $max, Min = $hist{$kmin}, Total = $count\n");

File: analysis/calc.lua
local lsqlite3 = require("lsqlite3")

local dbfile = "/home/elimtob/Workspace/mymemtrace/traces/imagick_r-test.db"
local db = lsqlite3.open(dbfile) or os.exit(1)

local rows = 0
for r in db:rows("SELECT SUM(REFS) from ROW_COUNT") do rows = r[1] end
print("Rows: " .. rows)

local hist   = {}
local max    = 0
local kmin   = 0
local count  = 0
local gcount = 0
local cl
hist[0] = math.maxinteger
for r in db:rows("SELECT ADDR, ADDR >> 6 FROM MEMREFS") do
    cl = r[2]
    local v = hist[cl]
    if not v then
        count = count + 1
        v = 0
    end
    v = v + 1
    hist[cl] = v
    if v > max        then max = v end
    if v < hist[kmin] then kmin = cl end
    gcount = gcount + 1
    if (gcount % 10000 == 0) then
        io.write("max = " .. max .. ", min = " .. hist[kmin] .. ", " .. gcount / rows * 100 .. "% done, .. v = " .. v .. ", cl = ".. cl .. "\r")
    end
end
print()
print("Max = " .. max .. ", Min = " .. (hist[kmin] or "nil") .. ", Total = " .. count )
db:close()
